File: Backend/app/pw_library/python_methods/crout.py
import numpy as np

def crout_decomposition(A):
    """
    Performs Crout decomposition on matrix A.
    Returns matrices L and U such that A = L * U.
    """
    n = A.shape[0]
    L = np.zeros((n, n), dtype=float)
    U = np.eye(n, dtype=float)  # U is an identity matrix

    for j in range(n):
        for i in range(j, n):
            # Compute L[i][j]
            L[i, j] = A[i, j] - sum(L[i, k] * U[k, j] for k in range(j))

        for i in range(j + 1, n):
            # Compute U[j][i]
            if L[j, j] == 0:
                raise ValueError("Matrix is singular; zero pivot encountered.")
            U[j, i] = (A[j, i] - sum(L[j, k] * U[k, i] for k in range(j))) / L[j, j]

    return L, U

def forward_substitution(L, b):
    """
    Solves the equation Ly = b for y, where L is a lower triangular matrix.
    """
    n = len(b)
    y = np.zeros_like(b, dtype=float)

    for i in range(n):
        y[i] = (b[i] - np.dot(L[i, :i], y[:i])) / L[i, i]

    return y

def backward_substitution(U, y):
    """
    Solves the equation Ux = y for x, where U is an upper triangular matrix.
    """
    n = len(y)
    x = np.zeros_like(y, dtype=float)

    for i in range(n - 1, -1, -1):
        x[i] = (y[i] - np.dot(U[i, i+1:], x[i+1:])) / U[i, i]

    return x

def solve_crout(A, b):
    """
    Solves the linear system Ax = b using Crout decomposition.
    """
    # Perform Crout decomposition
    L, U = crout_decomposition(A)

    # Solve Ly = b for y
    y = forward_substitution(L, b)

    # Solve Ux = y for x
    x = backward_substitution(U, y)

    return x, y, L, U

File: Backend/app/pw_library/rust_methods/crout.rs
use nalgebra::{DMatrix, DVector};

// Crout Decomposition
fn crout_decomposition(a: &DMatrix<f64>) -> Result<(DMatrix<f64>, DMatrix<f64>), &'static str> {
    let n = a.nrows();
    let mut l = DMatrix::zeros(n, n);
    let mut u = DMatrix::identity(n, n);

    for j in 0..n {
        for i in j..n {
            // Compute L[i][j]
            l[(i, j)] = a[(i, j)] - (0..j).map(|k| l[(i, k)] * u[(k, j)]).sum::<f64>();
        }

        for i in (j + 1)..n {
            // Compute U[j][i]
            if l[(j, j)] == 0.0 {
                return Err("Matrix is singular; zero pivot encountered.");
            }
            u[(j, i)] = (a[(j, i)] - (0..j).map(|k| l[(j, k)] * u[(k, i)]).sum::<f64>()) / l[(j, j)];
        }
    }

    Ok((l, u))
}

// Forward substitution for Ly = b
fn forward_substitution(l: &DMatrix<f64>, b: &DVector<f64>) -> DVector<f64> {
    let n = b.len();
    let mut y = DVector::zeros(n);

    for i in 0..n {
        let sum = l.row(i).columns(0, i).dot(&y.rows(0, i));
        y[i] = (b[i] - sum) / l[(i, i)];
    }

    y
}

// Backward substitution for Ux = y
fn backward_substitution(u: &DMatrix<f64>, y: &DVector<f64>) -> DVector<f64> {
    let n = y.len();
    let mut x = DVector::zeros(n);

    for i in (0..n).rev() {
        let sum = u.row(i).columns(i + 1, n - i - 1).dot(&x.rows(i + 1, n - i - 1));
        x[i] = (y[i] - sum) / u[(i, i)];
    }

    x
}

// Solving Ax = b using Crout decomposition
fn solve_crout(a: &DMatrix<f64>, b: &DVector<f64>) -> Result<(DVector<f64>, DVector<f64>, DMatrix<f64>, DMatrix<f64>), &'static str> {
    // Perform Crout decomposition
    let (l, u) = crout_decomposition(a)?;

    // Solve Ly = b for y
    let y = forward_substitution(&l, b);

    // Solve Ux = y for x
    let x = backward_substitution(&u, &y);

    Ok((x, y, l, u))
}
